File: apps/api/app/tools/google_ads.py
"""Google Ads Keyword Planner tool."""

import os
import tempfile
from typing import Any

from crewai.tools import tool

from app.tools.base import CredentialError, resolve_credential_value


def get_google_ads_client(credential: dict[str, Any]):
    """Get a Google Ads client with credentials.
    
    Args:
        credential: Credential document with type='google_ads'
        
    Returns:
        GoogleAdsClient instance
        
    Raises:
        CredentialError: If credentials are missing or invalid
    """
    try:
        from google.ads.googleads.client import GoogleAdsClient
    except ImportError:
        raise CredentialError(
            "google-ads package not installed. Run: pip install google-ads"
        )
    
    storage_method = credential.get("storageMethod", "env")
    
    def resolve(field: str) -> str:
        value = credential.get(field)
        if not value:
            raise CredentialError(f"Missing {field} in credential")
        if storage_method == "env":
            env_value = os.environ.get(value)
            if not env_value:
                raise CredentialError(
                    f"Environment variable '{value}' not set for {field}"
                )
            return env_value
        return value
    
    # Build config dict
    config = {
        "developer_token": resolve("googleAdsDeveloperToken"),
        "client_id": resolve("googleAdsClientId"),
        "client_secret": resolve("googleAdsClientSecret"),
        "refresh_token": resolve("googleAdsRefreshToken"),
        "login_customer_id": resolve("googleAdsCustomerId").replace("-", ""),
        "use_proto_plus": True,
    }
    
    # GoogleAdsClient.load_from_dict expects a specific format
    # We need to write a temp YAML file or use load_from_dict
    try:
        return GoogleAdsClient.load_from_dict(config)
    except Exception as e:
        # Try writing a temp YAML file as fallback
        yaml_content = f"""developer_token: {config['developer_token']}
client_id: {config['client_id']}
client_secret: {config['client_secret']}
refresh_token: {config['refresh_token']}
login_customer_id: {config['login_customer_id']}
use_proto_plus: True
"""
        with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml", delete=False) as f:
            f.write(yaml_content)
            temp_path = f.name
        
        try:
            return GoogleAdsClient.load_from_storage(temp_path)
        finally:
            os.unlink(temp_path)


@tool
def google_ads_keyword_ideas(query: str, credential: dict[str, Any]) -> str:
    """
    Fetch keyword ideas and search volume from Google Ads Keyword Planner API.
    Returns keyword suggestions with volume, competition, and bid estimates.
    
    Args:
        query: Seed keyword to generate ideas from
        credential: Google Ads credential document
        
    Returns:
        Keyword ideas with search volume, competition, and bid estimates
        
    Raises:
        CredentialError: If credentials are missing or invalid
    """
    client = get_google_ads_client(credential)
    
    storage_method = credential.get("storageMethod", "env")
    customer_id_field = credential.get("googleAdsCustomerId")
    
    if storage_method == "env":
        customer_id = os.environ.get(customer_id_field, "").replace("-", "")
    else:
        customer_id = customer_id_field.replace("-", "") if customer_id_field else ""
    
    if not customer_id:
        raise CredentialError("Could not resolve Google Ads customer ID")
    
    try:
        keyword_plan_idea_service = client.get_service("KeywordPlanIdeaService")
        
        request = client.get_type("GenerateKeywordIdeasRequest")
        request.customer_id = customer_id
        request.language = "languageConstants/1000"  # English
        request.geo_target_constants = ["geoTargetConstants/2840"]  # United States
        request.include_adult_keywords = False
        request.keyword_plan_network = client.enums.KeywordPlanNetworkEnum.GOOGLE_SEARCH
        
        request.keyword_seed.keywords.append(query)
        
        keyword_ideas = keyword_plan_idea_service.generate_keyword_ideas(request=request)
        
        result = f"""
GOOGLE ADS KEYWORD PLANNER DATA
===============================
Seed keyword: "{query}"

Keyword Ideas:
"""
        
        ideas_list = list(keyword_ideas)[:30]
        
        for idea in ideas_list:
            keyword = idea.text
            metrics = idea.keyword_idea_metrics
            
            avg_searches = metrics.avg_monthly_searches if metrics.avg_monthly_searches else 0
            competition = metrics.competition.name if metrics.competition else "UNKNOWN"
            low_bid = metrics.low_top_of_page_bid_micros / 1_000_000 if metrics.low_top_of_page_bid_micros else 0
            high_bid = metrics.high_top_of_page_bid_micros / 1_000_000 if metrics.high_top_of_page_bid_micros else 0
            
            result += f"""
Keyword: {keyword}
  Avg Monthly Searches: {avg_searches:,}
  Competition: {competition}
  Top of Page Bid: ${low_bid:.2f} - ${high_bid:.2f}
"""
        
        return result
        
    except Exception as e:
        raise CredentialError(f"Error fetching keyword data: {str(e)}")

File: packages/mastra-backend/src/mastra/tools/google-ads.ts
/**
 * Google Ads Keyword Planner tool.
 *
 * Uses the Google Ads REST API v18 directly — no npm package needed.
 * Requires a google_ads credential with:
 *   developerToken, clientId, clientSecret, refreshToken, customerId
 *
 * The credential is pre-injected by the runner via `resolveAgentTools`,
 * so the LLM only needs to provide the `query` parameter.
 */

import { createTool } from "@mastra/core/tools";
import { z } from "zod";
import { type CredentialConfig, resolveCredentialValue } from "../../sanity.js";

const GOOGLE_ADS_API_VERSION = "v19";
const TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token";

/**
 * Exchange a refresh token for a short-lived access token.
 */
async function getAccessToken(
  clientId: string,
  clientSecret: string,
  refreshToken: string
): Promise<string> {
  const resp = await fetch(TOKEN_ENDPOINT, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: clientId,
      client_secret: clientSecret,
      refresh_token: refreshToken,
      grant_type: "refresh_token",
    }),
    signal: AbortSignal.timeout(15_000),
  });

  if (!resp.ok) {
    const text = await resp.text();
    throw new Error(`Token refresh failed (${resp.status}): ${text}`);
  }

  const data: any = await resp.json();
  if (!data.access_token) {
    throw new Error(`No access_token in token response: ${JSON.stringify(data)}`);
  }
  return data.access_token;
}

/**
 * Call the Google Ads generateKeywordIdeas REST endpoint.
 */
async function generateKeywordIdeas(
  accessToken: string,
  developerToken: string,
  customerId: string,
  loginCustomerId: string,
  query: string
): Promise<any> {
  const url = `https://googleads.googleapis.com/${GOOGLE_ADS_API_VERSION}/customers/${customerId}:generateKeywordIdeas`;

  const headers: Record<string, string> = {
    Authorization: `Bearer ${accessToken}`,
    "developer-token": developerToken,
    "login-customer-id": loginCustomerId || customerId,
    "Content-Type": "application/json",
  };

  const body = {
    keywordSeed: {
      keywords: [query],
    },
    language: "languageConstants/1000", // English
    geoTargetConstants: ["geoTargetConstants/2840"], // United States
    keywordPlanNetwork: "GOOGLE_SEARCH",
  };

  const resp = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(30_000),
  });

  if (!resp.ok) {
    const text = await resp.text();
    // Strip HTML for cleaner error messages
    const clean = text.includes("<!DOCTYPE")
      ? `HTTP ${resp.status} — the Google Ads API may not be enabled in your GCP project. ` +
        `Enable it at: https://console.cloud.google.com/apis/library/googleads.googleapis.com ` +
        `(URL tried: ${url})`
      : `${resp.status}: ${text.slice(0, 500)}`;
    throw new Error(clean);
  }

  return resp.json();
}

/**
 * Format the API response into a readable string.
 */
function formatResults(query: string, data: any): string {
  const results = data.results ?? [];

  let output = `
GOOGLE ADS KEYWORD PLANNER DATA
===============================
Seed keyword: "${query}"
Language: English
Geo: United States

Keyword Ideas (${results.length}):
`;

  if (results.length === 0) {
    output += "\nNo keyword ideas returned for this query.\n";
    return output;
  }

  // Take top 30
  for (const idea of results.slice(0, 30)) {
    const keyword = idea.text ?? "";
    const metrics = idea.keywordIdeaMetrics ?? {};

    const avgSearches = metrics.avgMonthlySearches ?? 0;
    const competition = metrics.competition ?? "UNKNOWN";
    const lowBid = metrics.lowTopOfPageBidMicros
      ? (Number(metrics.lowTopOfPageBidMicros) / 1_000_000).toFixed(2)
      : "N/A";
    const highBid = metrics.highTopOfPageBidMicros
      ? (Number(metrics.highTopOfPageBidMicros) / 1_000_000).toFixed(2)
      : "N/A";

    output += `
Keyword: ${keyword}
  Avg Monthly Searches: ${Number(avgSearches).toLocaleString()}
  Competition: ${competition}
  Top of Page Bid: $${lowBid} - $${highBid}
`;
  }

  return output;
}

// ─── Mastra tool ──────────────────────────────────────────────

export const googleAdsKeywordIdeas = createTool({
  id: "google_ads_keyword_ideas",
  description:
    "Fetch keyword ideas and search volume from Google Ads Keyword Planner API. " +
    "Returns keyword suggestions with volume, competition, and bid estimates.",
  inputSchema: z.object({
    query: z.string().describe("Seed keyword to generate ideas from"),
    credentialJson: z
      .string()
      .describe("JSON-encoded Google Ads credential (auto-injected, do not set)"),
  }),
  execute: async ({ context: { query, credentialJson } }) => {
    try {
      const credential: CredentialConfig = JSON.parse(credentialJson);

      const developerToken = resolveCredentialValue(credential, "googleAdsDeveloperToken");
      const clientId = resolveCredentialValue(credential, "googleAdsClientId");
      const clientSecret = resolveCredentialValue(credential, "googleAdsClientSecret");
      const refreshToken = resolveCredentialValue(credential, "googleAdsRefreshToken");
      const rawCustomerId = resolveCredentialValue(credential, "googleAdsCustomerId");
      const customerId = rawCustomerId.replace(/-/g, "").trim();

      if (!customerId || !/^\d+$/.test(customerId)) {
        return `Google Ads error: Invalid customer ID "${rawCustomerId}". Expected a numeric ID like "123-456-7890" or "1234567890".`;
      }

      console.log(`[google_ads] Query: "${query}", Customer ID: ${customerId.slice(0, 4)}...`);

      // Get a fresh access token
      const accessToken = await getAccessToken(clientId, clientSecret, refreshToken);

      // Call the Keyword Planner API
      const data = await generateKeywordIdeas(
        accessToken,
        developerToken,
        customerId,
        customerId, // login_customer_id same as customer_id by default
        query
      );

      return formatResults(query, data);
    } catch (e: any) {
      return `Google Ads error: ${e.message}`;
    }
  },
});
